### PR TITLE
[ODS-5026] Create "Ed-Fi ODS Admin App" and "AB Connect" claimsets during the EdFi_Security database setup/ creation

### PR DESCRIPTION
**Pull request by tnunnerydl**

### CODE REVIEW

File: Artifacts/MsSql/Data/Security/2000-AdminApp-ClaimSets.sql [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+DECLARE @claimSetName  nvarchar(255)
+SET @claimSetName = 'Ed-Fi ODS Admin App'
+SET @applicationName = 'Ed-Fi ODS API'
+
+SELECT @applicationId = ApplicationId FROM  Applications WHERE ApplicationName = @applicationName
+
+PRINT 'Ensuring Ed-Fi ODS Admin App Claimset exists.'
+INSERT INTO ClaimSets (ClaimSetName, Application_ApplicationId)
+SELECT DISTINCT @claimSetName, @applicationId FROM ClaimSets
+WHERE NOT EXISTS (SELECT *
```
Comment (semalaiappan) [inline Artifacts/MsSql/Data/Security/2000-AdminApp-ClaimSets.sql:14]:
```suggestion
WHERE NOT EXISTS (SELECT 1
```

File: Artifacts/MsSql/Data/Security/2000-AdminApp-ClaimSets.sql [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+DECLARE @claimSetId int
+
+SELECT @actionId = ActionId FROM Actions WHERE ActionName = @actionName
+SELECT @claimSetId = ClaimSetId FROM ClaimSets WHERE ClaimSetName = @claimSetName
+
+PRINT 'Configuring Claims for Ed-Fi ODS Admin App Claimset...'
+INSERT INTO ClaimSetResourceClaims
+    (Action_ActionId, ClaimSet_ClaimSetId, ResourceClaim_ResourceClaimId, AuthorizationStrategyOverride_AuthorizationStrategyId)
+SELECT ActionId, @claimSetId, ResourceClaimId , @authorizationStrategyId FROM Actions a, @resourceClaimIds rc
+WHERE NOT EXISTS (SELECT *
```
Comment (semalaiappan) [inline Artifacts/MsSql/Data/Security/2000-AdminApp-ClaimSets.sql:47]:
```suggestion
WHERE NOT EXISTS (SELECT 1
```

File: Artifacts/MsSql/Data/Security/2000-AdminApp-ClaimSets.sql [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+DECLARE @claimSetName  nvarchar(255)
+SET @claimSetName = 'AB Connect'
+SET @applicationName = 'Ed-Fi ODS API'
+
+SELECT @applicationId = ApplicationId FROM  Applications WHERE ApplicationName = @applicationName
+
+PRINT 'Ensuring AB Connect Claimset exists.'
+INSERT INTO ClaimSets (ClaimSetName, Application_ApplicationId)
+SELECT DISTINCT @claimSetName, @applicationId FROM ClaimSets
+WHERE NOT EXISTS (SELECT *
```
Comment (semalaiappan) [inline Artifacts/MsSql/Data/Security/2000-AdminApp-ClaimSets.sql:72]:
```suggestion
WHERE NOT EXISTS (SELECT 1
```

File: Artifacts/MsSql/Data/Security/2000-AdminApp-ClaimSets.sql [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+DECLARE @claimSetId int
+
+SELECT @actionId = ActionId FROM Actions WHERE ActionName = @actionName
+SELECT @claimSetId = ClaimSetId FROM ClaimSets WHERE ClaimSetName = @claimSetName
+
+PRINT 'Configuring Claims for AB Connect Claimset...'
+INSERT INTO ClaimSetResourceClaims
+    (Action_ActionId, ClaimSet_ClaimSetId, ResourceClaim_ResourceClaimId)
+SELECT ActionId, @claimSetId, ResourceClaimId FROM Actions a, @resourceClaimIds rc
+WHERE NOT EXISTS (SELECT *
```
Comment (semalaiappan) [inline Artifacts/MsSql/Data/Security/2000-AdminApp-ClaimSets.sql:99]:
```suggestion
WHERE NOT EXISTS (SELECT 1
```

File: Artifacts/MsSql/Data/Security/2000-AdminApp-ClaimSets.sql [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+SELECT @claimSetId = ClaimsetId
+FROM dbo.ClaimSets c
+WHERE c.ClaimSetName = 'Assessment Vendor'
+
+PRINT 'Ensuring create and read actions for performanceLevelDescriptor are assigned to Assessment Vendor claimset'
+
+INSERT INTO ClaimSetResourceClaims
+    (Action_ActionId, ClaimSet_ClaimSetId, ResourceClaim_ResourceClaimId)
+SELECT ActionId, @claimSetId, @performanceLevelDescriptorClaimId FROM Actions a
+WHERE a.ActionName in ('Read','Create') AND NOT EXISTS (SELECT *
```
Comment (semalaiappan) [inline Artifacts/MsSql/Data/Security/2000-AdminApp-ClaimSets.sql:146]:
```suggestion
WHERE a.ActionName in ('Read','Create') AND NOT EXISTS (SELECT 1
```

**Review by semalaiappan — CHANGES_REQUESTED** [review]

File: Artifacts/MsSql/Data/Security/2000-AdminApp-ClaimSets.sql [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+
+DECLARE @actionName nvarchar(255)
+DECLARE @claimSetName nvarchar(255)
+DECLARE @resourceNames TABLE (ResourceName nvarchar(255))
+DECLARE @resourceClaimIds TABLE (ResourceClaimId int)
+DECLARE @authorizationStrategyId INT
+DECLARE @ResourceClaimId INT
+
+SET @claimSetName = 'Ed-Fi ODS Admin App'
+PRINT 'Creating temorary records.'
```
Comment (semalaiappan) [inline Artifacts/MsSql/Data/Security/2000-AdminApp-ClaimSets.sql:29]:
```suggestion
PRINT 'Creating Temporary Records.'
```

File: Artifacts/MsSql/Data/Security/2000-AdminApp-ClaimSets.sql [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+
+-- Configure AB Connect ClaimSet
+
+DECLARE @actionName nvarchar(255)
+DECLARE @claimSetName nvarchar(255)
+DECLARE @resourceNames TABLE (ResourceName nvarchar(255))
+DECLARE @resourceClaimIds TABLE (ResourceClaimId int)
+
+SET @claimSetName = 'AB Connect'
+PRINT 'Creating temorary records.'
```
Comment (semalaiappan) [inline Artifacts/MsSql/Data/Security/2000-AdminApp-ClaimSets.sql:85]:
```suggestion
PRINT 'Creating Temporary Records.'
```

**Review by semalaiappan — CHANGES_REQUESTED** [review]

File: Artifacts/MsSql/Data/Security/2000-AdminApp-ClaimSets.sql [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+DECLARE @applicationId int
+DECLARE @applicationName  nvarchar(max)
+DECLARE @claimSetName  nvarchar(255)
+SET @claimSetName = 'Ed-Fi ODS Admin App'
+SET @applicationName = 'Ed-Fi ODS API'
+
+SELECT @applicationId = ApplicationId FROM  Applications WHERE ApplicationName = @applicationName
+
+PRINT 'Ensuring Ed-Fi ODS Admin App Claimset exists.'
+INSERT INTO ClaimSets (ClaimSetName, Application_ApplicationId)
```
Comment (semalaiappan) [inline Artifacts/MsSql/Data/Security/2000-AdminApp-ClaimSets.sql:12]:
```suggestion
INSERT INTO dbo.ClaimSets (ClaimSetName, Application_ApplicationId)
```

File: Artifacts/MsSql/Data/Security/2000-AdminApp-ClaimSets.sql [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+INSERT INTO @resourceClaimIds SELECT ResourceClaimId FROM ResourceClaims WHERE ResourceName IN (SELECT ResourceName FROM @resourceNames)
+
+SELECT @authorizationStrategyId = AuthorizationStrategyId
+FROM   AuthorizationStrategies
+WHERE  AuthorizationStrategyName = 'NoFurtherAuthorizationRequired'
+
+DECLARE @actionId int
+DECLARE @claimSetId int
+
+SELECT @actionId = ActionId FROM Actions WHERE ActionName = @actionName
```
Comment (semalaiappan) [inline Artifacts/MsSql/Data/Security/2000-AdminApp-ClaimSets.sql:40]: tables requires to have dbo. before the table name . Please check

**Review by semalaiappan — APPROVED** [review]

**Review by semalaiappan — CHANGES_REQUESTED** [review]
Formatting is required for Sql scripts . Please check other scripts here for reference Ed-Fi-ODS\Artifacts\MsSql\Data\Security

File: Artifacts/MsSql/Data/Security/2000-AdminApp-ClaimSets.sql [hunk 10]
```diff
@@ -0,0 +1,150 @@
+-- Create Ed-Fi ODS Admin App ClaimSet
+
+DECLARE @applicationId int
+DECLARE @applicationName  nvarchar(max)
+DECLARE @claimSetName  nvarchar(255)
+SET @claimSetName = 'Ed-Fi ODS Admin App'
+SET @applicationName = 'Ed-Fi ODS API'
+
+SELECT @applicationId = ApplicationId FROM  Applications WHERE ApplicationName = @applicationName
```
Comment (semalaiappan) [inline Artifacts/MsSql/Data/Security/2000-AdminApp-ClaimSets.sql:9]:
```suggestion
SELECT @applicationId = ApplicationId FROM  dbo.Applications WHERE ApplicationName = @applicationName
```

File: Artifacts/MsSql/Data/Security/2000-AdminApp-ClaimSets.sql [hunk 11]
```diff
@@ -0,0 +1,150 @@
...
+
+PRINT 'Ensuring Ed-Fi ODS Admin App Claimset exists.'
+INSERT INTO ClaimSets (ClaimSetName, Application_ApplicationId)
+SELECT DISTINCT @claimSetName, @applicationId FROM ClaimSets
```
Comment (semalaiappan) [inline Artifacts/MsSql/Data/Security/2000-AdminApp-ClaimSets.sql:13]:
```suggestion
SELECT DISTINCT @claimSetName, @applicationId FROM dbo.ClaimSets
```

File: Artifacts/MsSql/Data/Security/2000-AdminApp-ClaimSets.sql [hunk 12]
```diff
@@ -0,0 +1,150 @@
...
+WHERE NOT EXISTS (SELECT 1
+                  FROM ClaimSets
```
Comment (semalaiappan) [inline Artifacts/MsSql/Data/Security/2000-AdminApp-ClaimSets.sql:15]:
```suggestion
                  FROM dbo.ClaimSets
```

File: Artifacts/MsSql/Data/Security/2000-AdminApp-ClaimSets.sql (hunk 9 above)
Comment (semalaiappan) [inline Artifacts/MsSql/Data/Security/2000-AdminApp-ClaimSets.sql:40]:
```suggestion
SELECT @actionId = ActionId FROM dbo.Actions WHERE ActionName = @actionName
```

File: Artifacts/MsSql/Data/Security/2000-AdminApp-ClaimSets.sql [hunk 13]
```diff
@@ -0,0 +1,150 @@
...
+WHERE  AuthorizationStrategyName = 'NoFurtherAuthorizationRequired'
+
+DECLARE @actionId int
+DECLARE @claimSetId int
+
+SELECT @actionId = ActionId FROM Actions WHERE ActionName = @actionName
+SELECT @claimSetId = ClaimSetId FROM ClaimSets WHERE ClaimSetName = @claimSetName
+
+PRINT 'Configuring Claims for Ed-Fi ODS Admin App Claimset...'
+INSERT INTO ClaimSetResourceClaims
```
Comment (semalaiappan) [inline Artifacts/MsSql/Data/Security/2000-AdminApp-ClaimSets.sql:44]:
```suggestion
INSERT INTO dbo.ClaimSetResourceClaims
```

File: Artifacts/MsSql/Data/Security/2000-AdminApp-ClaimSets.sql [hunk 14]
```diff
@@ -0,0 +1,150 @@
...
+    (Action_ActionId, ClaimSet_ClaimSetId, ResourceClaim_ResourceClaimId, AuthorizationStrategyOverride_AuthorizationStrategyId)
+SELECT ActionId, @claimSetId, ResourceClaimId , @authorizationStrategyId FROM Actions a, @resourceClaimIds rc
```
Comment (semalaiappan) [inline Artifacts/MsSql/Data/Security/2000-AdminApp-ClaimSets.sql:46]:
```suggestion
SELECT ActionId, @claimSetId, ResourceClaimId , @authorizationStrategyId FROM dbo.Actions a, @resourceClaimIds rc
```

File: Artifacts/MsSql/Data/Security/2000-AdminApp-ClaimSets.sql [hunk 15]
```diff
@@ -0,0 +1,150 @@
...
+WHERE NOT EXISTS (SELECT 1
+                  FROM ClaimSetResourceClaims
+				  WHERE Action_ActionId = a.ActionId AND ClaimSet_ClaimSetId = @claimSetId AND ResourceClaimId = rc.ResourceClaimId)
+
+SELECT @actionId = ActionId FROM Actions WHERE ActionName = 'Read'
```
Comment (semalaiappan) [inline Artifacts/MsSql/Data/Security/2000-AdminApp-ClaimSets.sql:51]:
```suggestion
SELECT @actionId = ActionId FROM dbo.Actions WHERE ActionName = 'Read'
```

File: Artifacts/MsSql/Data/Security/2000-AdminApp-ClaimSets.sql [hunk 16]
```diff
@@ -0,0 +1,150 @@
...
+SELECT @ResourceClaimId = ResourceClaimId FROM ResourceClaims WHERE ResourceName = 'types'
```
Comment (semalaiappan) [inline Artifacts/MsSql/Data/Security/2000-AdminApp-ClaimSets.sql:52]:
```suggestion
SELECT @ResourceClaimId = ResourceClaimId FROM dbo.ResourceClaims WHERE ResourceName = 'types'
```

File: Artifacts/MsSql/Data/Security/2000-AdminApp-ClaimSets.sql [hunk 17]
```diff
@@ -0,0 +1,150 @@
...
+INSERT INTO ClaimSetResourceClaims
```
Comment (semalaiappan) [inline Artifacts/MsSql/Data/Security/2000-AdminApp-ClaimSets.sql:53]:
```suggestion
INSERT INTO dbo.ClaimSetResourceClaims
```

File: Artifacts/MsSql/Data/Security/2000-AdminApp-ClaimSets.sql [hunk 18]
```diff
@@ -0,0 +1,150 @@
...
+DECLARE @applicationId int
+DECLARE @applicationName  nvarchar(max)
+DECLARE @claimSetName  nvarchar(255)
+SET @claimSetName = 'AB Connect'
+SET @applicationName = 'Ed-Fi ODS API'
+
+SELECT @applicationId = ApplicationId FROM  Applications WHERE ApplicationName = @applicationName
+
+PRINT 'Ensuring AB Connect Claimset exists.'
+INSERT INTO ClaimSets (ClaimSetName, Application_ApplicationId)
```
Comment (semalaiappan) [inline Artifacts/MsSql/Data/Security/2000-AdminApp-ClaimSets.sql:70]:
```suggestion
INSERT INTO dbo.ClaimSets (ClaimSetName, Application_ApplicationId)
```

File: Artifacts/MsSql/Data/Security/2000-AdminApp-ClaimSets.sql [hunk 19]
```diff
@@ -0,0 +1,150 @@
...
+SELECT DISTINCT @claimSetName, @applicationId FROM ClaimSets
+WHERE NOT EXISTS (SELECT 1
+                  FROM ClaimSets
```
Comment (semalaiappan) [inline Artifacts/MsSql/Data/Security/2000-AdminApp-ClaimSets.sql:73]:
```suggestion
                  FROM dbo.ClaimSets
```

File: Artifacts/MsSql/Data/Security/2000-AdminApp-ClaimSets.sql [hunk 20]
```diff
@@ -0,0 +1,150 @@
...
+
+DECLARE @actionName nvarchar(255)
+DECLARE @claimSetName nvarchar(255)
+DECLARE @resourceNames TABLE (ResourceName nvarchar(255))
+DECLARE @resourceClaimIds TABLE (ResourceClaimId int)
+
+SET @claimSetName = 'AB Connect'
+PRINT 'Creating Temporary Records.'
+INSERT INTO @resourceNames VALUES ('gradeLevelDescriptor'),('academicSubjectDescriptor'),('publicationStatusDescriptor'),('educationStandards')
+INSERT INTO @resourceClaimIds SELECT ResourceClaimId FROM ResourceClaims WHERE ResourceName IN (SELECT ResourceName FROM @resourceNames)
```
Comment (semalaiappan) [inline Artifacts/MsSql/Data/Security/2000-AdminApp-ClaimSets.sql:87]:
```suggestion
INSERT INTO @resourceClaimIds SELECT ResourceClaimId FROM dbo.ResourceClaims WHERE ResourceName IN (SELECT ResourceName FROM @resourceNames)
```

File: Artifacts/MsSql/Data/Security/2000-AdminApp-ClaimSets.sql [hunk 21]
```diff
@@ -0,0 +1,150 @@
...
+
+DECLARE @actionId int
+DECLARE @claimSetId int
+
+SELECT @actionId = ActionId FROM Actions WHERE ActionName = @actionName
```
Comment (semalaiappan) [inline Artifacts/MsSql/Data/Security/2000-AdminApp-ClaimSets.sql:92]:
```suggestion
SELECT @actionId = ActionId FROM dbo.Actions WHERE ActionName = @actionName
```

File: Artifacts/MsSql/Data/Security/2000-AdminApp-ClaimSets.sql [hunk 22]
```diff
@@ -0,0 +1,150 @@
...
+SELECT @claimSetId = ClaimSetId FROM ClaimSets WHERE ClaimSetName = @claimSetName
```
Comment (semalaiappan) [inline Artifacts/MsSql/Data/Security/2000-AdminApp-ClaimSets.sql:93]:
```suggestion
SELECT @claimSetId = ClaimSetId FROM dbo.ClaimSets WHERE ClaimSetName = @claimSetName
```

File: Artifacts/MsSql/Data/Security/2000-AdminApp-ClaimSets.sql [hunk 23]
```diff
@@ -0,0 +1,150 @@
...
+
+PRINT 'Configuring Claims for AB Connect Claimset...'
+INSERT INTO ClaimSetResourceClaims
```
Comment (semalaiappan) [inline Artifacts/MsSql/Data/Security/2000-AdminApp-ClaimSets.sql:96]:
```suggestion
INSERT INTO dbo.ClaimSetResourceClaims
```

**Review by semalaiappan — CHANGES_REQUESTED** [review]

**Comment by vimayya** [comment]
@tbmeador-dlp, @semalaiappan my understanding is that these scripts need to be idempotent. I am seeing additional records being inserted on ClaimSetResourceClaims when run more than once.

**Review by semalaiappan — APPROVED** [review]

**Pull request merged**